### ISSUE TITLE
AsyncLocal<> does not play nice with IAsyncDisposable interface

### ISSUE THREAD

**Issue by ibasin** [dotnet/coreclr#24748]
AsyncLocal which is supposed to represent "ambient data that is local to a given asynchronous control flow, such as an asynchronous method" does not work when IAsyncDisposable interface is used. Changes to the Value of an AsyncLocal<> that are made inside the DisposeAsync() method do not persist outside of the DisposeAsync() method meaning that they are counted as running is a separate logical context.

The issue is easy to reproduce. Create a new .NET Core Console app and replace the content of Program.cs with the code in the attached Program.txt. When you run it, the expected result is 0 1 0 0, while the program produces 0 1 0 1 (because changes made in DisposeAsync() do not persist). 

This is a significant bug that prevents us using the ambient content pattern. I see no workaround as CallContext is not supported in .Net Core. 

Please help!

P.S. I am using VS2019 Preview and .NET Core 3.0 Preview5

Ilya Basin
262-893-9741

[Program.txt](https://github.com/dotnet/coreclr/files/3214283/Program.txt)

**Comment by jkotas** [comment]
cc @stephentoub

**Comment by stephentoub** [comment]
This doesn't have anything to do with IAsyncDisposable: it has to do with methods implemented using async/await (which you've used to implement your DisposeAsync method).  When you invoke an `async` method, a guard is erected that prevents any changes to ExecutionContext from flowing out from the async method to the caller.  That's been the behavior since async/await were introduced.

**Comment by ibasin** [comment]
That makes sense. Thank you.

What do you think would be a good way of incrementing on creation and decrementing on disposal of an ambient context variable if a copy of the ExecutionContext is created when DisposeAsync() is called?

**Comment by stephentoub** [comment]
> What do you think would be a good way of incrementing on creation and decrementing on disposal of an ambient context variable if a new copy of the ExecutionContext is created when DisposeAsync() is called and discarded when it is done?

What are you actually trying to achieve? i.e. can you share a small sample of code that would be using this type?  What is meant to happen when the count hits zero?

> I see no workaround as CallContext is not supported in .Net Core.

CallContext would behave the same way; it's stored in ExecutionContext as well.  Are you porting something from .NET Framework that was working there?

**Comment by ibasin** [comment]
I am building EF Core data access with unit of work that stores its context in teh ambient variable and save the data on when using statements scope ends. Units of work in the ambient context are stored in the stack so that they can be nested. I figured out how to make it work. I just wrapped my stack in a wrapper that derives from MarshalByRefObject like this:

        private sealed class StackWrapper : MarshalByRefObject
        {
            public ImmutableStack<IDataContext> Stack { get; set; } = ImmutableStack.Create<IDataContext>();
        }

This seems to solve all of my problems. Thank you for quick response!
